File: pkg/pairing/cpace.go
// Package pairing provides CPace (Composable PAKE) for secure device pairing.
// CPace is defined in RFC 9497 and provides security against offline dictionary attacks.
//
// Unlike the old implementation, attackers cannot brute-force passwords offline
// even if they capture the entire exchange. They must participate online for each guess.
package pairing

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"fmt"
	"io"

	"github.com/ivere27/nitella/pkg/crypto"
	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"
)

// GetSessionID returns the session ID for coordination between peers
func (c *CPaceSession) GetSessionID() []byte {
	return c.sessionID
}

const (
	ScalarSize = 32
	PointSize  = 32
	keySize    = 32 // internal, use KeySize from pake.go for public API
)

// CPaceSession represents one side of a CPace key exchange.
// CPace provides security against offline dictionary attacks.
type CPaceSession struct {
	role       string
	sessionID  []byte // Shared session identifier (e.g., pairing code)
	myScalar   []byte // Random private scalar (ya or yb)
	myPublic   []byte // Public value (Ya or Yb)
	peerPublic []byte // Peer's public value
	generator  []byte // Password-derived generator G'
	sharedKey  []byte // Final session key
}

// NewCPaceSession creates a new CPace session.
// role: "cli" or "node"
// password: shared secret (pairing code)
func NewCPaceSession(role string, password []byte) (*CPaceSession, error) {
	return NewCPaceSessionWithID(role, password, nil)
}

// NewCPaceSessionWithID creates a new CPace session with an explicit session ID.
// role: "cli" or "node"
// password: shared secret (pairing code)
// sessionID: unique session identifier (if nil, derived from password for compatibility)
// Using distinct sessionID strengthens transcript binding per RFC 9497
func NewCPaceSessionWithID(role string, password, sessionID []byte) (*CPaceSession, error) {
	if role != RoleCLI && role != RoleNode {
		return nil, fmt.Errorf("role must be '%s' or '%s'", RoleCLI, RoleNode)
	}

	// Derive sessionID from password if not provided (ensures both parties have same ID)
	if sessionID == nil {
		h := sha256.Sum256(append([]byte("cpace-session-id:"), password...))
		sessionID = h[:16]
	}

	// Derive generator: G' = Hash(password || sessionID || context) * Basepoint
	generator := deriveGenerator(password, sessionID)

	// Generate random scalar
	scalar := make([]byte, ScalarSize)
	if _, err := io.ReadFull(rand.Reader, scalar); err != nil {
		return nil, fmt.Errorf("failed to generate random scalar: %w", err)
	}
	clampScalar(scalar)

	// Compute public value: Y = scalar * G'
	public, err := curve25519.X25519(scalar, generator)
	if err != nil {
		return nil, fmt.Errorf("failed to compute public value: %w", err)
	}

	return &CPaceSession{
		role:      role,
		sessionID: sessionID,
		myScalar:  scalar,
		myPublic:  public,
		generator: generator,
	}, nil
}

// deriveGenerator derives a curve point from password using hash-and-multiply.
// G' = H(password || sessionID || context) * Basepoint
func deriveGenerator(password, sessionID []byte) []byte {
	// Domain separation for security
	h := sha256.New()
	h.Write([]byte("cpace-v1-curve25519"))
	h.Write(password)
	h.Write(sessionID)
	digest := h.Sum(nil)

	// Clamp for X25519 scalar
	clampScalar(digest)

	// G' = digest * Basepoint
	generator, _ := curve25519.X25519(digest, curve25519.Basepoint)

	// Wipe password-derived scalar
	crypto.Wipe(digest)

	return generator
}

// clampScalar clamps a scalar for X25519 as per RFC 7748
func clampScalar(s []byte) {
	s[0] &= 248
	s[31] &= 127
	s[31] |= 64
}

// GetPublicValue returns the public value to send to peer.
func (c *CPaceSession) GetPublicValue() []byte {
	return c.myPublic
}

// SetPeerPublic sets the peer's public value and derives the shared key.
func (c *CPaceSession) SetPeerPublic(peerPublic []byte) error {
	if len(peerPublic) != PointSize {
		return fmt.Errorf("invalid peer public value size: %d", len(peerPublic))
	}

	// Check for low-order points (all zeros = identity)
	allZero := true
	for _, b := range peerPublic {
		if b != 0 {
			allZero = false
			break
		}
	}
	if allZero {
		return fmt.Errorf("invalid peer public value: identity point")
	}

	c.peerPublic = peerPublic

	// Compute shared secret: K = myScalar * peerPublic
	sharedSecret, err := curve25519.X25519(c.myScalar, peerPublic)
	if err != nil {
		return fmt.Errorf("failed to compute shared secret: %w", err)
	}

	// Derive session key: key = HKDF(K, Ya || Yb || sessionID)
	c.sharedKey = c.deriveSessionKey(sharedSecret)

	// Wipe shared secret after key derivation
	crypto.Wipe(sharedSecret)

	return nil
}

// deriveSessionKey derives the final session key from shared secret and transcript.
func (c *CPaceSession) deriveSessionKey(sharedSecret []byte) []byte {
	// Order public values deterministically (CLI first)
	var transcript []byte
	if c.role == RoleCLI {
		transcript = append(transcript, c.myPublic...)
		transcript = append(transcript, c.peerPublic...)
	} else {
		transcript = append(transcript, c.peerPublic...)
		transcript = append(transcript, c.myPublic...)
	}
	transcript = append(transcript, c.sessionID...)

	// HKDF with transcript as info
	hkdfReader := hkdf.New(sha256.New, sharedSecret, nil, transcript)
	key := make([]byte, keySize)
	io.ReadFull(hkdfReader, key)
	return key
}

// GetSharedKey returns the derived shared key (nil if exchange not complete).
func (c *CPaceSession) GetSharedKey() []byte {
	return c.sharedKey
}

// IsComplete returns true if the key exchange is complete.
func (c *CPaceSession) IsComplete() bool {
	return c.sharedKey != nil
}

// Encrypt encrypts data with the shared key using AES-GCM.
func (c *CPaceSession) Encrypt(plaintext []byte) (ciphertext, nonce []byte, err error) {
	if c.sharedKey == nil {
		return nil, nil, fmt.Errorf("key exchange not complete")
	}

	block, err := aes.NewCipher(c.sharedKey)
	if err != nil {
		return nil, nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, nil, err
	}

	nonce = make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, nil, err
	}

	ciphertext = gcm.Seal(nil, nonce, plaintext, nil)
	return ciphertext, nonce, nil
}

// Decrypt decrypts data with the shared key using AES-GCM.
func (c *CPaceSession) Decrypt(ciphertext, nonce []byte) ([]byte, error) {
	if c.sharedKey == nil {
		return nil, fmt.Errorf("key exchange not complete")
	}

	block, err := aes.NewCipher(c.sharedKey)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	return gcm.Open(nil, nonce, ciphertext, nil)
}

// DeriveConfirmationEmoji derives emoji fingerprint from shared key for visual confirmation.
func (c *CPaceSession) DeriveConfirmationEmoji() string {
	if c.sharedKey == nil {
		return ""
	}

	hash := sha256.Sum256(c.sharedKey)

	emojis := []string{
		"🐶", "🐱", "🐭", "🐹", "🐰", "🦊", "🐻", "🐼",
		"🐨", "🐯", "🦁", "🐮", "🐷", "🐸", "🐵", "🐔",
		"🐧", "🐦", "🐤", "🦆", "🦅", "🦉", "🦇", "🐺",
		"🐗", "🐴", "🦄", "🐝", "🐛", "🦋", "🐌", "🐞",
		"🌸", "🌺", "🌻", "🌹", "🌷", "🌼", "🌿", "🍀",
		"🍎", "🍊", "🍋", "🍇", "🍓", "🍒", "🍑", "🥝",
		"🌙", "⭐", "🌟", "✨", "⚡", "🔥", "🌈", "☀️",
		"🎸", "🎹", "🎺", "🎷", "🥁", "🎻", "🎤", "🎧",
	}

	result := ""
	for i := 0; i < 4; i++ {
		idx := int(hash[i*2]) % len(emojis)
		result += emojis[idx]
	}

	return result
}

// Close securely wipes all sensitive key material from the session.
// Should be called when the session is no longer needed.
func (c *CPaceSession) Close() {
	crypto.Wipe(c.myScalar)
	crypto.Wipe(c.generator)
	crypto.Wipe(c.sharedKey)
	c.myScalar = nil
	c.generator = nil
	c.sharedKey = nil
}

File: nitellad-rs/src/cpace.rs
use aes_gcm::{aead::Aead, Aes256Gcm, KeyInit, Nonce};
use anyhow::{bail, Context, Result};
use hkdf::Hkdf;
use rand::RngCore;
use sha2::{Digest, Sha256};
use x25519_dalek::{PublicKey, StaticSecret};

pub const ROLE_CLI: &str = "cli";
pub const ROLE_NODE: &str = "node";

pub struct CPaceSession {
    role: String,
    session_id: Vec<u8>,
    my_secret: StaticSecret,
    my_public: PublicKey,
    peer_public: Option<PublicKey>,
    shared_key: Option<Vec<u8>>,
}

impl CPaceSession {
    pub fn new(role: &str, password: &[u8], session_id_opt: Option<&[u8]>) -> Result<Self> {
        if role != ROLE_CLI && role != ROLE_NODE {
            bail!("Invalid role");
        }

        // Derive session ID if not provided
        let session_id = match session_id_opt {
            Some(id) => id.to_vec(),
            None => {
                let mut hasher = Sha256::new();
                hasher.update(b"cpace-session-id:");
                hasher.update(password);
                let result = hasher.finalize();
                result[..16].to_vec()
            }
        };

        // Derive Generator G'
        let generator_point = derive_generator(password, &session_id);

        // Generate random scalar (my_secret)
        let mut rng = rand::thread_rng();
        let my_secret = StaticSecret::random_from_rng(&mut rng);

        // Compute Public Value Y = my_secret * G'
        // In x25519-dalek, diffie_hellman(secret, point) = secret * point
        let my_public = my_secret.diffie_hellman(&generator_point);
        let my_public_point = PublicKey::from(my_public.to_bytes());

        Ok(Self {
            role: role.to_string(),
            session_id,
            my_secret,
            my_public: my_public_point,
            peer_public: None,
            shared_key: None,
        })
    }

    pub fn get_public_value(&self) -> [u8; 32] {
        *self.my_public.as_bytes()
    }

    pub fn set_peer_public(&mut self, peer_bytes: &[u8]) -> Result<()> {
        if peer_bytes.len() != 32 {
            bail!("Invalid peer public length");
        }
        let mut arr = [0u8; 32];
        arr.copy_from_slice(peer_bytes);
        let peer_public = PublicKey::from(arr);

        // Check for identity/low order? x25519-dalek handles some checks, but RFC 7748 allows all 32-byte strings.
        // Go implementation checks for all-zero.
        if peer_bytes.iter().all(|&b| b == 0) {
            bail!("Invalid peer public: identity point");
        }

        // Compute K = my_secret * peer_public
        let k = self.my_secret.diffie_hellman(&peer_public);

        // Derive Session Key
        let key = self.derive_session_key(k.as_bytes(), &peer_public);
        self.peer_public = Some(peer_public);
        self.shared_key = Some(key);

        Ok(())
    }

    fn derive_session_key(&self, shared_secret: &[u8], peer_public: &PublicKey) -> Vec<u8> {
        // Transcript: (CLI_Pub || Node_Pub || SessionID)
        let mut transcript = Vec::new();
        if self.role == ROLE_CLI {
            transcript.extend_from_slice(self.my_public.as_bytes());
            transcript.extend_from_slice(peer_public.as_bytes());
        } else {
            transcript.extend_from_slice(peer_public.as_bytes());
            transcript.extend_from_slice(self.my_public.as_bytes());
        }
        transcript.extend_from_slice(&self.session_id);

        // Match Go: hkdf.New(sha256.New, sharedSecret, nil, transcript)
        // = HKDF(IKM=sharedSecret, salt=nil, info=transcript)
        let hkdf = Hkdf::<Sha256>::new(None, shared_secret);
        let mut okm = [0u8; 32];
        hkdf.expand(&transcript, &mut okm)
            .expect("HKDF expand failed");
        okm.to_vec()
    }

    pub fn derive_confirmation_emoji(&self) -> String {
        let shared_key = match &self.shared_key {
            Some(k) => k,
            None => return String::new(),
        };

        let hash = Sha256::digest(shared_key);

        // Must match Go's cpace.go emoji list exactly
        const EMOJIS: &[&str] = &[
            "🐶", "🐱", "🐭", "🐹", "🐰", "🦊", "🐻", "🐼", "🐨", "🐯", "🦁", "🐮", "🐷", "🐸",
            "🐵", "🐔", "🐧", "🐦", "🐤", "🦆", "🦅", "🦉", "🦇", "🐺", "🐗", "🐴", "🦄", "🐝",
            "🐛", "🦋", "🐌", "🐞", "🌸", "🌺", "🌻", "🌹", "🌷", "🌼", "🌿", "🍀", "🍎", "🍊",
            "🍋", "🍇", "🍓", "🍒", "🍑", "🥝", "🌙", "⭐", "🌟", "✨", "⚡", "🔥", "🌈", "☀️",
            "🎸", "🎹", "🎺", "🎷", "🥁", "🎻", "🎤", "🎧",
        ];

        let mut result = String::new();
        for i in 0..4 {
            let idx = (hash[i * 2] as usize) % EMOJIS.len();
            result.push_str(EMOJIS[idx]);
        }
        result
    }

    pub fn encrypt(&self, plaintext: &[u8]) -> Result<(Vec<u8>, Vec<u8>)> {
        let key_bytes = self.shared_key.as_ref().context("Handshake not complete")?;
        let key = aes_gcm::Key::<Aes256Gcm>::from_slice(key_bytes);
        let cipher = Aes256Gcm::new(key);

        let mut nonce_bytes = [0u8; 12];
        rand::thread_rng().fill_bytes(&mut nonce_bytes);
        let nonce = Nonce::from_slice(&nonce_bytes);

        let ciphertext = cipher
            .encrypt(nonce, plaintext)
            .map_err(|e| anyhow::anyhow!("Encryption failed: {}", e))?;
        Ok((ciphertext, nonce_bytes.to_vec()))
    }

    pub fn decrypt(&self, ciphertext: &[u8], nonce_bytes: &[u8]) -> Result<Vec<u8>> {
        let key_bytes = self.shared_key.as_ref().context("Handshake not complete")?;
        let key = aes_gcm::Key::<Aes256Gcm>::from_slice(key_bytes);
        let cipher = Aes256Gcm::new(key);
        let nonce = Nonce::from_slice(nonce_bytes);

        let plaintext = cipher
            .decrypt(nonce, ciphertext)
            .map_err(|e| anyhow::anyhow!("Decryption failed: {}", e))?;
        Ok(plaintext)
    }
}

fn derive_generator(password: &[u8], session_id: &[u8]) -> PublicKey {
    // H(password || sessionID || context)
    let mut hasher = Sha256::new();
    hasher.update(b"cpace-v1-curve25519");
    hasher.update(password);
    hasher.update(session_id);
    let digest = hasher.finalize();

    // In x25519, we don't have a direct "scalar * basepoint" where basepoint is arbitrary point in PublicKey form?
    // Wait, standard X25519 is "Scalar * Basepoint(9)".
    // To implement "Scalar * G'", we treat G' as a "Public Key" (Point) and Scalar as "Private Key".
    //
    // Go logic:
    // clamp(digest) -> scalar
    // G' = scalar * Basepoint

    let mut scalar_bytes = [0u8; 32];
    scalar_bytes.copy_from_slice(&digest);
    // Clamp
    scalar_bytes[0] &= 248;
    scalar_bytes[31] &= 127;
    scalar_bytes[31] |= 64;

    let secret = StaticSecret::from(scalar_bytes);
    let public = PublicKey::from(&secret);
    public
}
